### PR TITLE
Json read and write tests

### PR DESCRIPTION
**Pull request by deepaksa1**
This PR aims to add read and write tests for json data type.

### CODE REVIEW

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+                string jsonData = reader.GetString(0);
+                _output.WriteLine(jsonData);
+                Assert.NotNull(jsonData);
+            }
+            System.Collections.ObjectModel.ReadOnlyCollection<DbColumn> schema = reader.GetColumnSchema();
+            foreach (DbColumn column in schema)
+            {
+                _output.WriteLine("Column Name is " + column.ColumnName);
+                _output.WriteLine("Column DataType is " + column?.DataType.ToString());
+                _output.WriteLine("Column DataTypeName is " + column.DataTypeName);
```
Comment (saurabh500) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs:72]: You may want to assert that the datatype is json

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    public class JsonTest
+    {
+        private readonly ITestOutputHelper _output;
+
+        public JsonTest(ITestOutputHelper output)
+        {
+            _output = output;
+        }
+
+        [ConditionalFact(typeof(DataTestUtility), nameof(DataTestUtility.IsJsonSupported))]
```
Comment (deepaksa1) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs:22]: @David-Engel @cheenamalhotra 

Need some help on the usage of `ConditionalFact` 
I am trying to set this test to run only if JSON is supported. However, this test also runs even if `IsJsonSupported` is set as `false`

Comment (JRahnama) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs:22]: @deepaksa1Are you using Console or Visual Studio? `xunit.runner.visualstudio` and `xunit.runner.console` work differently. In Visual Studio, you might not be able to run platform-specific tests or certain xUnit extensions as expected. It's better to test using the Console to see the accurate outcome. (This is by my experience. @David-Engel or @cheenamalhotra may have better inputs to help you on this)

Comment (David-Engel) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs:22]: JRahnama is correct. I've seen VS run tests even when the ConditionalFact says it shouldn't. Console results should behave correctly.

Comment (deepaksa1) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs:22]: Thanks @JRahnama and @David-Engel for pointing it out. I ran the tests from console and the results were as expected.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs (hunk 1 above)
Comment (deepaksa1) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs]: added

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        public void TestJsonWrite()
+        {
+            AppContext.SetSwitch("Switch.Microsoft.Data.SqlClient.UseManagedNetworkingOnWindows", true);
```
Comment (saurabh500) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs:26]: Why ? 
Don't use this AppContext switch.

**Comment by ErikEJ** [comment]
How about reader.GetFieldValue<JsonDocument> (which together  with SqlDdType.Json for SqlParameter) is waht EF Core needs. I see no tests for that? (But I could be looking in the wrong place)

**Comment by saurabh500** [comment]
@ErikEJ the support hasn't been added yet. The API is on the list of things to add.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+                    //Write json value using SP
+                    using (SqlCommand command2 = connection.CreateCommand())
+                    {
+                        command2.CommandText = spName;
+                        command2.CommandType = CommandType.StoredProcedure;
+                        var parameter2 = new SqlParameter("@jsonData", SqlDbTypeExtensions.Json);
+                        parameter2.Value = jsonDataString;
+                        command2.Parameters.Add(parameter2);
+                        int rowsAffected3 = command2.ExecuteNonQuery();
+                        _output.WriteLine($"Rows affected: {rowsAffected3}");
```
Comment (saurabh500) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs:77]: No asserts?

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+                    }
+                    reader.Close();
+
+                    //Test 3
+                    //Read json value using SP
+                    using (SqlCommand command2 = connection.CreateCommand())
+                    {
+                        command2.CommandText = spName;
+                        command2.CommandType = CommandType.StoredProcedure;
+                        SqlDataReader reader2 = command2.ExecuteReader();
```
Comment (saurabh500) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs:148]: You could try wrapping this in using block

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+                    //Test 2
+                    //Read the column metadata
+                    System.Collections.ObjectModel.ReadOnlyCollection<DbColumn> schema = reader.GetColumnSchema();
+                    foreach (DbColumn column in schema)
+                    {
+                        _output.WriteLine("Column Name is " + column.ColumnName);
+                        _output.WriteLine("Column DataType is " + column?.DataType.ToString());
+                        _output.WriteLine("Column DataTypeName is " + column.DataTypeName);
+                        Assert.Equal("json",column.DataTypeName);
```
Comment (saurabh500) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs:138]: Space after ,

Comment (saurabh500) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs:138]: Can you format this code? Looks like a few locations could use space after comma

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs (hunk 3 above)
Comment (deepaksa1) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs]: removed

File: src/Microsoft.Data.SqlClient/tests/tools/Microsoft.Data.SqlClient.TestUtilities/Config.cs [hunk 7]
```diff
@@ -42,6 +42,7 @@ public class Config
         public string KerberosDomainUser = null;
         public bool IsManagedInstance = false;
         public string AliasName = null;
+        public bool IsJsonSupported = false;
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/tests/tools/Microsoft.Data.SqlClient.TestUtilities/Config.cs:45]: Is this something you can detect after making connection?

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+    {
+        private readonly ITestOutputHelper _output;
+
+        public JsonTest(ITestOutputHelper output)
+        {
+            _output = output;
+        }
+
+        [ConditionalFact(typeof(DataTestUtility), nameof(DataTestUtility.IsJsonSupported))]
+        public void TestJsonWrite()
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs:24]: Please add async API testing as well.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+                    using (SqlCommand command2 = connection.CreateCommand())
+                    {
+                        command2.CommandText = spName;
+                        command2.CommandType = CommandType.StoredProcedure;
+                        SqlDataReader reader2 = command2.ExecuteReader();
+                        while (reader2.Read())
+                        {
+                            string jsonData = reader2.GetString(0);
+                            _output.WriteLine(jsonData);
+                            Assert.NotNull(jsonData);
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs:149]: I don't think this is a very good way to validate, can you compare jsonData with source data that was written to DB?

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+                    SqlDataReader reader = command.ExecuteReader();
+                    while (reader.Read())
+                    {
+                        string jsonData = reader.GetString(0);
+                        _output.WriteLine(jsonData);
+                        Assert.Equal(jsonDataString,jsonData);
+                    }
+
+                    //Test 2
+                    //Read the column metadata
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs:127]: I'd recommend adding multiple tests instead of running all tests in 1 big one. You can create tables in setup and drop them in tear down.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs (hunk 4 above)
Comment (deepaksa1) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs:77]: Added asserts for write tests

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs (hunk 5 above)
Comment (deepaksa1) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs]: Added `using` block

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs (hunk 6 above)
Comment (deepaksa1) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs]: Done.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs (hunk 9 above)
Comment (deepaksa1) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs]: Modified the assert statement

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs (hunk 10 above)
Comment (deepaksa1) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs:130]: These are quite small tests (single operation) and uses common sql command. Should I break them into individual tests?

File: src/Microsoft.Data.SqlClient/tests/tools/Microsoft.Data.SqlClient.TestUtilities/Config.cs (hunk 7 above)
Comment (saurabh500) [reply src/Microsoft.Data.SqlClient/tests/tools/Microsoft.Data.SqlClient.TestUtilities/Config.cs:45]: @cheenamalhotra 
Can't be done without turning on Feature Switches on Debug builds of Sql server today. 
We will need a hardcoded config value at this moment unfortunately :( 

also question for my education @cheenamalhotra . In ConditionalFact check, do we make server calls to check for features. Because that may lead to transient fault handling in test execution condition. I like the idea of dynamically checking the server for the feature, as long as it is not done in the body of the test, since the test would show as executed but it really wont be executed. So adding a server query in ConditionalFact condition would make sense. But wanted to check if there is prior art.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs (hunk 8 above)
Comment (saurabh500) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs:24]: @deepaksa1  this is great feedback.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs (hunk 10 above)
Comment (saurabh500) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs:130]: I think it makes sense to keep this test as is. Data is being read and various aspects of the data are being verified, i.e. Data / Metadata.

However it would be nicer to see these refactored into various methods.
e.g. You could have a reader being executed and data read/asserted in a different function, and then from the same reader extract the Schema and send it to another function to validate. 

Having multiple tests also tend to increase runtime, for setup which, could otherwise have been done in a single test. There is no golden rule. But for this test, it makes sense to have a single `ConditionalFact` and then have multiple verification methods being called from the same test.

Basically 
```
var reader = command.ExecuteReader()
ValidateRows(reader);
ValidateScehma(reader.GetColumnSchema());
reader.Close();
```

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs (hunk 8 above)
Comment (deepaksa1) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs:65]: Added async tests

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs (hunk 10 above)
Comment (deepaksa1) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/JsonTest/JsonTest.cs:215]: Refactored test cases.
